// gws/ui/file.tsx
import * as React from 'react';

import * as base from './base';
import * as util from './util';

import {Touchable} from './button';

export interface FileInputProps extends base.InputProps<FileList> {
    multiple?: boolean;
    accept?: string;
    listFormat?: (files: FileList) => React.ReactNode,
}


export class FileInput extends base.Control<FileInputProps> {

    inputRef: React.RefObject<HTMLInputElement>;

    constructor(props) {
        super(props);
        this.inputRef = React.createRef();
    }

    render() {
        let input = null;

        if (!this.props.disabled) {
            input = <input
                type="file"
                accept={this.props.accept}
                disabled={this.props.disabled}
                multiple={this.props.multiple}
                onChange={e => this.onChange(e)}
                ref={this.inputRef}
                style={{display: 'none'}}
            />;
        }

        let hasFiles = this.props.value && this.props.value.length > 0;
        let cls = util.className('uiFileInput', hasFiles && 'uiHasContent')

        let listFormat = this.props.listFormat || defaultListFormat;

        return <base.Content of={this} withClass={cls}>
            <base.Box>
                {input}
                <button
                    className="uiRawButton"
                    disabled={this.props.disabled}
                    ref={this.focusRef}
                    onClick={() => this.onClick()}
                    title={this.props.tooltip || ''}
                />
                {listFormat(this.props.value)}
            </base.Box>
        </base.Content>
    }

    protected onClick() {
        this.grabFocus();
        this.inputRef.current.click();
    }

    protected onChange(e: React.SyntheticEvent<any>) {
        if (this.props.whenChanged)
            this.props.whenChanged(e.currentTarget.files);
    }
}

function formatFileSize(size) {
    if (size >= 1e9) {
        return (size / 1e9).toFixed(2) + ' GB';
    }
    if (size >= 1e6) {
        return (size / 1e6).toFixed(1) + ' MB';
    }
    if (size >= 1e3) {
        return (size / 1e3).toFixed(0) + ' KB';
    }
    return size.toFixed(0) + ' B';
}

function defaultListFormat(fileList) {
    if (!fileList || !fileList.length)
        return null;

    let fs = [].slice.call(fileList, 0).map(f => {
        return f.name + ' (' + formatFileSize(f.size) + ')';
    })

    return <div className='uiFileInputList'>{fs.join(', ')}</div>;
}
