// gws/ui/file.tsx
import * as React from 'react';

import * as base from './base';
import * as util from './util';

import {Touchable} from './button';

export interface FileInputProps extends base.InputProps<FileList> {
    multiple?: boolean;
    accept?: string;
    listFormat?: (files: FileList) => React.ReactNode,
}


export class FileInput extends base.Control<FileInputProps> {

    inputRef: React.RefObject<HTMLInputElement>;

    constructor(props) {
        super(props);
        this.inputRef = React.createRef();
    }

    render() {
        let input = null;

        if (!this.props.disabled) {
            input = <input
                type="file"
                accept={this.props.accept}
                disabled={this.props.disabled}
                multiple={this.props.multiple}
                onChange={e => this.onChange(e)}
                ref={this.inputRef}
                style={{display: 'none'}}
            />;
        }

        let hasFiles = this.props.value && this.props.value.length > 0;
        let cls = util.className('uiFileInput', hasFiles && 'uiHasFiles')

        let listFormat = this.props.listFormat || defaultListFormat;

        return <base.Content of={this} withClass={cls}>
            <base.Box>
                {input}
                    <button
                        className="uiRawButton"
                        disabled={this.props.disabled}
                        ref={this.focusRef}
                        onClick={() => this.onClick()}
                    />
                {listFormat(this.props.value)}
            </base.Box>
        </base.Content>
    }

    protected onClick() {
        this.grabFocus();
        this.inputRef.current.click();
    }

    protected onChange(e: React.SyntheticEvent<any>) {
        if (this.props.whenChanged)
            this.props.whenChanged(e.currentTarget.files);
    }
}

function defaultListFormat(fileList) {
    if (!fileList || !fileList.length)
        return null;

    let fs = [].slice.call(fileList, 0).map(f => {
        let name = f.name.slice(0, 50),
            size = (f.size / 1024).toFixed(2) + 'K';
        return name + ' (' + size + ')';
    })

    return <div className='uiFileInputList'>{fs.join(', ')}</div>;
}
