// gws/ui/text.tsx
import * as React from 'react';

import {UiPureComponent, UiProps} from './base';

interface ErrorProps extends UiProps {
    text: string;
    longText?: string;
    fade?: boolean;
}

interface TextProps extends UiProps {
    content: string;
    whenTouched?: () => any;
    className?: string;
}

interface LinkProps extends TextProps {
    href?: string;
    target?: string;
}

interface TextBlockProps extends UiProps {
    content: string;
    withHTML?: boolean;
    className?: string;
    whenTouched?: () => any;
}

interface HtmlBlockProps extends UiProps {
    content: string;
    className?: string;
}

export class Error extends UiPureComponent<ErrorProps> {
    ref: React.RefObject<HTMLDivElement>;

    componentDidMount() {
        if(this.props.fade)
            setTimeout(() => this.ref.current && (this.ref.current.className += ' isHidden'));
    }

    render() {
        this.ref = React.createRef();
        return <div ref={this.ref} className='uiError'>
            <div className="uiErrorText">{this.props.text}</div>
            {this.props.longText && <div className="uiErrorLongText">{this.props.longText}</div>}
        </div>
    }
}

export class Title extends UiPureComponent<TextProps> {
    render() {
        let cls = this.props.className || 'uiTitle';
        return <Text {...this.props} className={cls}/>;
    }
}

export class Text extends UiPureComponent<TextProps> {
    render() {
        return <div
            className={this.props.className || 'uiText'}
            onClick={this.props.whenTouched}
        >{this.props.content}</div>;
    }
}

export class Link extends UiPureComponent<LinkProps> {
    render() {
        let cls = this.props.className || 'uiLink';

        if (this.props.whenTouched) {
            return <a className={cls} onClick={() => this.props.whenTouched()}>{this.props.content}</a>;
        }
        if (this.props.href) {
            return <a className={cls} href={this.props.href}>{this.props.content}</a>;
        }
        return <a className={cls}>{this.props.content}</a>;
    }
}

export class TextBlock extends UiPureComponent<TextBlockProps> {

    render() {
        let s = String(this.props.content || '').trim();
        if (!s)
            return null;

        let cls = this.props.className || 'uiTextBlock';
        let c;

        c = (s[0] === '<' && this.props.withHTML) ? asHtml(s) : s;
        return <div
            className={cls}
            onClick={this.props.whenTouched}
        >{c}</div>;
    }
}

export class HtmlBlock extends UiPureComponent<HtmlBlockProps> {
    render() {
        let s = String(this.props.content || '').trim();
        if (!s)
            return null;
        return <div
            className={this.props.className || 'uiTextBlock'}
        >{asHtml(s)}</div>;
    }
}

function asHtml(s) {
    return <div dangerouslySetInnerHTML={{__html: s}}/>;
}
