// gws/types.ts
import * as React from 'react';
import * as ol from 'openlayers';
import * as api from './core/gws-api';

export interface Dict {
    [key: string]: any;
}

export interface IStoreWrapper {
    addHook(kind, type, handler);
    connect(klass, props);
    dispatch(type, args);
    getValue(key);
    update(args);
    wrap(element);
}

export interface IApplication {
    domNode: HTMLDivElement;
    map: IMapManager;
    overviewMap: IMapManager;
    options: Dict;
    project: api.ProjectProps;
    server: api.GwsServerApi;
    store: IStoreWrapper;
    tags: Dict;

    createController(klass: any, parent: IController, cfg?: Dict);
    createControllerFromConfig(parent: IController, cfg: Dict);
    controller(uid: string): IController;

    initState(args: Dict);
    label(key);
    reload();

    whenRequestStarted(url);
    whenRequestEnded(url);

    whenMounted();
    whenLoaded();

    whenChanged(prop: string, fn: (value: any) => void);

    updateLocation(data: object);

    addTool(name: string, obj: ITool);
    startTool(name: string, temp?: boolean);
    stopTool(name: string);
}

export interface IController {
    uid: string;
    tag: string;
    app: IApplication;
    map: IMapManager;
    options: Dict;
    children: Array<IController>;
    parent?: IController;
    defaultView: React.ReactElement<any>;
    mapOverlayView: React.ReactElement<any>;
    appOverlayView: React.ReactElement<any>;

    init();
    update(args: any);
    touched();
    renderChildren(): Array<React.ReactElement<any>>;
    __(key: string): string;
}

export interface ITool extends IController {
    start();
    stop();
}

export interface IUser {
    displayName: string;
}

export interface IMapLayer {
    type: string;
    uid: string;
    title: string;
    attributes: Dict;
    extent: ol.Extent;

    attribution: string;
    description: string;

    parent?: IMapLayer;
    children: Array<IMapLayer>;
    hasChildren: boolean;

    expanded: boolean;
    visible: boolean;
    selected: boolean;
    listed: boolean;
    editable: boolean;
    unfolded: boolean;

    shouldDraw: boolean;
    shouldList: boolean;

    isSystem: boolean;

    oLayer?: ol.layer.Base;
    printItem?: api.PrintItem;

    show();
    hide();
    changed();
    beforeDraw();
    reset();
}

export interface IMapFeatureLayer extends IMapLayer {
    editStyle?: IMapStyle;
    features: Array<IMapFeature>;
    geometryType: string;
    source: ol.source.Vector;
    style?: IMapStyle;

    addFeature(feature: IMapFeature): boolean;
    addFeatures(features: Array<IMapFeature>): number;
    removeFeature(feature: IMapFeature);

    clear();
    setStyle(style: IMapStyle);

}

export interface MapViewState {
    centerX: number;
    centerY: number;
    resolution: number;
    scale: number;
    rotation: number;
    angle: number;
}

export interface IMapStyle {
    oStyle: ol.style.Style;
    props: api.StyleProps;
    getText(copy: boolean): ol.style.Text | null;
    getImage(copy: boolean): ol.style.Image | null;
    apply(geom: ol.geom.Geometry, label: string, resolution: number): Array<ol.style.Style>;
}

export interface DrawInteractionOptions {
    geometryType: string;
    geometryFunction?: any;
    style?: IMapStyle;
    whenEnded?: (olf: ol.Feature) => void;
    whenStarted?: (olf: ol.Feature) => void;
}

export interface ModifyInteractionOptions {
    style?: IMapStyle;
    source?: ol.source.Vector;
    features?: ol.Collection<ol.Feature>;
    whenEnded?: (olf: Array<ol.Feature>) => void;
    whenStarted?: (olfs: Array<ol.Feature>) => void;
}

export interface IMapManager {
    app: IApplication;
    bbox: ol.Extent;
    viewExtent: ol.Extent;
    crs: string;
    domNode: HTMLDivElement;
    extent: ol.Extent;
    oMap: ol.Map;
    oView: ol.View;
    projection: ol.proj.Projection;
    resolutions: Array<number>;
    root: IMapLayer;
    size: ol.Size;
    viewState: MapViewState;

    init(props: api.MapProps, appLoc: object);
    update(args: any);
    changed();

    addLayer(layer: IMapLayer, where: number, parent: IMapLayer);
    addTopLayer(layer: IMapLayer);
    addServiceLayer(layer: IMapFeatureLayer);

    removeLayer(layer: IMapLayer);
    getLayer(uid: string): IMapLayer;

    setLayerVisible(layer: IMapLayer, on: boolean);
    setLayerExpanded(layer: IMapLayer, on: boolean);
    hideAllLayers();
    deselectAllLayers();

    selectLayer(layer: IMapLayer);
    //queryLayerDescription(layer: IMapLayer);

    setResolution(n: number, animate?: boolean);
    setNextResolution(delta: number, animate?: boolean);
    setScale(n: number, animate?: boolean);
    setRotation(n: number, animate?: boolean);
    setAngle(n: number, animate?: boolean);
    setViewExtent(extent: ol.Extent, animate?: boolean);
    setCenter(c: ol.Coordinate, animate?: boolean);
    setViewState(vs: any, animate?: boolean);
    resetViewState(animate?: boolean);

    setInteracting(on: boolean);

    setInteractions(intrs: Array<ol.interaction.Interaction | string>);
    pushInteractions();
    popInteractions();

    drawInteraction(opts: DrawInteractionOptions);
    modifyInteraction(opts: ModifyInteractionOptions);
    pointerInteraction(opts: Dict);

    readFeature(fs: api.FeatureProps): IMapFeature;
    readFeatures(fs: Array<api.FeatureProps>): Array<IMapFeature>;
    writeFeatures(fs: Array<IMapFeature>): Array<api.FeatureProps>;

    getStyleFromSelector(selector: string): IMapStyle | null;
    getStyle(props: api.StyleProps): IMapStyle | null;
    styleFunction(style?: IMapStyle);

    geom2shape(geom: ol.geom.Geometry): api.ShapeProps;
    shape2geom(shape: api.ShapeProps) : ol.geom.Geometry;

    bitmapAsDataURL(boxRect: ClientRect): string;
    printParams(boxRect: ClientRect | null, dpi: number): Promise<api.PrintParams>;
    searchParams(keyword: string, geometry: ol.geom.Geometry | null): Promise<api.SearchParams>;

    formatCoordinate(n: number): string;
}

export interface IMapFeature {
    geometry?: ol.geom.Geometry;
    label: string;
    oFeature?: ol.Feature;
    printProps: api.PrintFeatureProps;
    props: api.FeatureProps;
    style?: IMapStyle;
    uid: string;

    setStyle(style: IMapStyle);
    setGeometry(geom: ol.geom.Geometry);
    setLabel(label: string);
}

export interface IMapFeatureArgs {
    props?: api.FeatureProps;
    geometry?: ol.geom.Geometry;
    oFeature?: ol.Feature;
    style?: IMapStyle;
    label?: string;
}




export interface ViewProps {
    controller: IController;
}

export interface ISidebarItem extends IController {
    iconClass: string;
    tooltip: string;
    tabView: React.ReactElement<any>;
}
