// gws/map/manager.ts
import * as ol from 'openlayers';
import * as proj4 from 'proj4';

import * as types from '../types';
import * as api from '../core/gws-api';
import * as tools from '../tools';

import * as layer from './layer';

let layerTypes = {
    'box': layer.BoxLayer,
    'group': layer.GroupLayer,
    'leaf': layer.LeafLayer,
    'root': layer.RootLayer,
    'tile': layer.TileLayer,
    'tree': layer.TreeLayer,
    'vector': layer.VectorLayer,
    'osm': layer.OsmLayer,
    'xyz': layer.XYZLayer,
};

import {Feature} from './feature';

import * as interactions from './interactions';

const POINTER_UPDATE_INTERVAL = 10;
const VIEW_UPDATE_INTERVAL = 1000;

const ANIMATION_DURATION = 600;
const ANIMATION_EASING = ol.easing.easeOut;

const USE_RAW_BITMAPS_FOR_PRINT = false;
const BITMAP_PRINT_DPI_THRESHOLD = 90.7; // matches OGC_SCREEN_PPI on the server

export class MapManager implements types.IMapManager {
    static STOP_WALK = {};

    app: types.IApplication = null;

    crs = '';
    domNode = null;
    oMap: ol.Map = null;
    oView: ol.View = null;
    projection = null;
    root = null;
    resolutions = null;
    extent = null;

    protected connectedToStore = false;
    protected coordinatePrecision = 0;
    protected defaultViewState = null;
    protected updateCount = 0;
    protected intrStack = [];
    protected standardInteractions = {};
    protected props: api.MapProps;

    // @TODO this should be 'viewExtent' actually
    get bbox() {
        return this.oView.calculateExtent();
    }

    get viewExtent() {
        return this.oView.calculateExtent();
    }

    get size() {
        return this.oMap.getSize();
    }

    get viewState(): types.MapViewState {
        if (this.oView) {
            let v = this.oView,
                c = v.getCenter(),
                res = v.getResolution(),
                rot = v.getRotation();

            return {
                centerX: c[0],
                centerY: c[1],
                resolution: res,
                scale: tools.res2scale(res),
                rotation: rot,
                angle: tools.rad2deg(rot),
            }
        }
    }

    get style() {
        return this.app.style;
    }

    constructor(app, connectedToStore) {
        this.app = app;
        this.connectedToStore = connectedToStore;
    }

    update(args) {
        if (this.connectedToStore)
            this.app.store.update(args);
    }

    async init(props, appLoc) {
        console.log('MAP_INIT', appLoc);

        this.props = props;
        this.domNode = document.querySelector('.gwsMap');

        this.resolutions = this.props.resolutions;
        this.extent = this.props.extent;
        this.coordinatePrecision = this.props.coordinatePrecision || 0;

        this.defaultViewState = {
            centerX: this.props.center[0],
            centerY: this.props.center[1],
            resolution: this.props.initResolution,
            rotation: 0
        };

        let vs = this.decodeViewState(appLoc['map']);

        this.initOMap(vs || this.defaultViewState);
        this.initLayers();
        this.initEvents();
        this.initInteractions();

        this.setLayerChecked(this.root, true);

        this.changed();

        if (this.connectedToStore)
            this.app.whenChanged('appLocation', loc => {
                let vs = this.decodeViewState(loc['map']);
                vs = vs || this.defaultViewState;
                console.log('vs from location', vs);
                this.setViewState(vs, true);
            });

        return await true;
    }

    changed() {
        if (this.connectedToStore) {
            this.update({
                mapUpdateCount: ++this.updateCount,
                mapRawUpdateCount: ++this.rawUpdateCount,
                mapAttribution: this.visibleAttributions(),
            });
        }
        this.draw();
    }

    forceUpdate() {
        this.walk(this.root, la => {
            if (la.oLayer && la.oLayer.getSource())
                la.oLayer.getSource().changed()
        });
    }


    //

    addLayer(layer, where, parent) {
        this.insertLayer(layer, where, parent);
        this.changed();
    }

    addTopLayer(layer) {
        this.addLayer(layer, 0, this.root);
    }

    addServiceLayer(layer: types.IMapFeatureLayer) {
        layer.visible = true;
        layer.listed = false;
        this.addTopLayer(layer);
        return layer;
    }

    removeLayer(layer) {
        if (layer && layer.parent)
            layer.parent.children = layer.parent.children.filter(la => la !== layer);
        this.changed();
    }

    getLayer(uid) {
        let found = null;

        this.walk(this.root, la => {
            if (la.uid === uid) {
                found = la;
                return MapManager.STOP_WALK;
            }
        });

        return found;
    }

    editableLayers() {
        return this.collect(this.root, la => la.editAccess);
    }

    setLayerChecked(layer, on) {
        function exclusiveCheck(la: types.IMapLayer, parents) {
            if (la.exclusive) {
                // in an exclusive group, if a layer belongs to the currenty checked path
                // it should be checked and others should be unchecked
                if (la.children.some(c => parents.indexOf(c) >= 0)) {
                    la.children.forEach(c => c.checked = parents.indexOf(c) >= 0);
                }
            }
            la.children.forEach(c => exclusiveCheck(c, parents));
        }

        function update(la, visible) {
            la.visible = visible && la.checked;
            la.children.forEach(c => update(c, la.visible));
        }

        if (!on) {
            layer.checked = false;
        } else {
            // check all the parent layers
            let parents = [];
            let la = layer;

            while (la) {
                la.checked = true;
                parents.push(la);
                la = la.parent;
            }
            // maintain exclusive groups
            exclusiveCheck(this.root, parents);
        }

        update(this.root, true);
        this.changed();
    }

    setLayerExpanded(layer, on) {
        layer.expanded = on;
        this.changed();
    }

    async selectLayer(layer) {
        if (!layer.description) {
            let res = await this.app.server.mapDescribeLayer({layerUid: layer.uid})
            layer.description = res.description;
        }
        this.walk(this.root, la => la.selected = false);
        layer.selected = true;
        this.update({
            'mapSelectedLayer': layer
        });
        this.changed();
    }

    deselectAllLayers() {
        this.walk(this.root, la => la.selected = false);
        this.update({
            'mapSelectedLayer': null
        });
        this.changed();
    }

    hideAllLayers() {
        this.walk(this.root, la => {
            if (!la.isSystem)
                la.visible = false
        });
        this.changed();
    }

    setTargetDomNode(node) {
        this.oMap.setTarget(node);
        this.changed();
    }

    //

    constrainScale(scale: number): number {
        let res = tools.clamp(
            tools.scale2res(scale),
            Math.min(...this.resolutions),
            Math.max(...this.resolutions),
        );
        return tools.res2scale(res);
    }

    protected prepareViewState(vs: any) {
        let p: any = {};

        if ('scale' in vs) {
            vs.resolution = tools.scale2res(vs.scale);
        }

        if ('resolution' in vs) {
            p.resolution = tools.clamp(
                tools.asNumber(vs.resolution),
                Math.min(...this.resolutions),
                Math.max(...this.resolutions),
            );
        }

        if ('center' in vs) {
            vs.centerX = vs.center[0];
            vs.centerY = vs.center[1];
        }

        if ('centerX' in vs && 'centerY' in vs) {
            p.center = this.oView.constrainCenter([
                tools.asNumber(vs.centerX),
                tools.asNumber(vs.centerY)
            ]);
        }

        if ('angle' in vs) {
            vs.rotation = tools.deg2rad(tools.asNumber(vs.angle));
        }

        if ('rotation' in vs) {
            let r = tools.clamp(tools.asNumber(vs.rotation), 0, Math.PI * 2);
            p.rotation = this.oView.constrainRotation(r);
        }

        return p;
    }

    setViewState(vs: any, animate: boolean) {
        let p = this.prepareViewState(vs);
        this.oView.cancelAnimations();

        console.log('setViewState, p', p);

        if (animate) {
            return this.oView.animate({
                ...p,
                duration: ANIMATION_DURATION,
                easing: ANIMATION_EASING
            }, () => this.viewChanged('animation'));
        }

        if ('center' in p)
            this.oView.setCenter(p.center);
        if ('resolution' in p)
            this.oView.setResolution(p.resolution);
        if ('rotation' in p)
            this.oView.setRotation(p.rotation);
    }

    setResolution(n, animate) {
        this.setViewState({resolution: n}, animate);
    }

    setScale(n, animate) {
        this.setViewState({scale: n}, animate);
    }

    setNextResolution(delta, animate) {
        let cur = this.viewState.resolution,
            res = (delta > 0)
                ? Math.min(...this.resolutions.filter(r => r > cur))
                : Math.max(...this.resolutions.filter(r => r < cur));

        if (Number.isFinite(res) && res !== cur)
            this.setResolution(res, animate);
    }

    setRotation(n, animate) {
        this.setViewState({rotation: n}, animate);
    }

    setAngle(n, animate) {
        this.setViewState({angle: n}, animate);
    }

    setCenter(c, animate) {
        this.setViewState({center: c}, animate);
    }

    setViewExtent(extent, animate, padding) {
        if (!extent)
            return;

        // view.fit doesn't have maxResolution, so compute the stuff here
        // @TODO: rotation

        padding = padding || 0;

        let size = this.oMap.getSize();
        let res = this.oView.getResolutionForExtent(extent, [
            size[0] - padding * 2, size[1] - padding * 2
        ]);

        this.setViewState({
            center: ol.extent.getCenter(extent),
            resolution: tools.clamp(res,
                Math.min(...this.resolutions),
                Math.max(...this.resolutions))
        }, animate);
    }

    resetViewState(animate) {
        this.setViewState(this.defaultViewState, animate);
    }

    //

    lockInteractions() {
        this.oMap.getInteractions().forEach(ix => ix.setActive(false));
    }

    unlockInteractions() {
        this.oMap.getInteractions().forEach(ix => ix.setActive(true));
    }

    setInteractions(ixs) {
        this.oMap.getInteractions().forEach(intr => intr.setActive(false));
        this.oMap.getInteractions().clear();
        ixs.forEach(intr => {
            if (typeof (intr) === 'string')
                intr = this.standardInteractions[intr];
            this.oMap.addInteraction(intr)
        });
        this.oMap.getInteractions().forEach(intr => intr.setActive(true));
    }

    defaultInteractions = ['DragPan', 'MouseWheelZoom', 'PinchZoom', 'ZoomBox'];

    appendInteractions(ixs) {
        this.setInteractions(this.defaultInteractions.concat(ixs.filter(Boolean)));
    }

    prependInteractions(ixs) {
        this.setInteractions(ixs.filter(Boolean).concat(this.defaultInteractions));
    }

    resetInteractions() {
        this.setInteractions(this.defaultInteractions);
    }

    pushInteractions() {
        let intrs = [];
        this.oMap.getInteractions().forEach(intr => {
            intr.setActive(false);
            intrs.push(intr);
        });
        this.intrStack.push(intrs);
    }

    popInteractions() {
        if (this.intrStack.length) {
            this.setInteractions(this.intrStack.pop());
        }
    }

    drawInteraction(opts) {
        return interactions.draw(this, opts);
    }

    selectInteraction(opts) {
        return interactions.select(this, opts);
    }

    modifyInteraction(opts) {
        return interactions.modify(this, opts);
    }

    snapInteraction(opts) {
        return interactions.snap(this, opts);
    }

    pointerInteraction(opts) {
        return interactions.pointer(this, opts);
    }

    //

    protected initOMap(vs) {
        console.log('initOMap', vs);

        vs = this.oView ? this.viewState : vs;

        this.crs = this.props.crs;

        if (this.crs === 'EPSG:3857') {
            this.projection = ol.proj.get(this.crs);
        } else {
            proj4.defs(this.crs, this.props.crsDef);
            this.projection = ol.proj.get(this.crs);
        }

        if (this.oMap) {
            this.oMap.setLayerGroup(new ol.layer.Group());
            this.oMap = this.oView = null;
        }

        this.oMap = new ol.Map({
            view: new ol.View({
                center: [vs.centerX, vs.centerY],
                resolution: vs.resolution,
                rotation: vs.rotation,
                resolutions: this.resolutions,
                extent: this.extent,
                projection: this.projection,
            }),
            // @TODO: need to pass window.devicePixelRatio around
            pixelRatio: 1,
            controls: [],
            interactions: [],

        });

        this.oView = this.oMap.getView();
    }

    protected initLayers() {

        this.root = this.initLayer({
            type: 'root',
            uid: '_root',
            title: this.props.title,
            meta: null,

            layers: this.props.layers,

            options: {
                expanded: true,
                visible: true,
                selected: false,
                listed: true,
            },
        });
    }

    protected initLayer(props: api.LayerProps, parent = null): types.IMapLayer {
        let cls = layerTypes[props.type];
        if (!cls)
            throw new Error('unknown layer type: ' + props.type);

        let layer = new cls(this, props);

        if (props['layers']) {
            props['layers'].forEach(p => {
                // // propagate group visibility down...
                // if (!props.options.visible)
                //     p.options.visible = false;
                this.initLayer(p, layer)
            });
        }

        if (parent)
            this.insertLayer(layer, -1, parent);

        return layer;
    }

    protected updateViewState() {
        if (this.connectedToStore) {

            let vs = this.viewState;
            this.update({
                mapUpdateCount: ++this.updateCount,
                mapViewState: vs,
                mapCenterX: vs.centerX,
                mapCenterY: vs.centerY,
                mapResolution: vs.resolution,
                mapRotation: vs.rotation,
                mapScale: vs.scale,
                mapAngle: vs.angle,
            });
        }
    }

    protected updateAppLocation() {
        if (this.connectedToStore) {
            let v1 = this.encodeViewState(this.viewState);
            let v2 = this.encodeViewState(this.defaultViewState);
            this.app.updateLocation({map: v1 === v2 ? null : v1});
        }
    }

    formatCoordinate(n) {
        return tools.toFixedMax(n, this.coordinatePrecision)

    }

    protected encodeViewState(vs) {
        let xs = [
            this.formatCoordinate(vs.centerX),
            this.formatCoordinate(vs.centerY),
            tools.res2scale(vs.resolution),
        ];

        if (vs.rotation) {
            xs.push(tools.rad2deg(vs.rotation))
        }

        return xs.join(',');
    }

    protected decodeViewState(h) {
        h = String(h || '').split(',');

        let xs = [
            parseFloat(h[0]),
            parseFloat(h[1]),
            parseInt(h[2])
        ];

        xs.push(h.length === 4 ? parseInt(h[3]) : 0);

        if (xs.some(x => Number.isNaN(x)))
            return null;

        return {
            centerX: xs[0],
            centerY: xs[1],
            resolution: tools.scale2res(xs[2]),
            rotation: tools.deg2rad(xs[3]),
        };
    }

    updatePointerPosition(cc) {
        this.update({
            mapPointerX: cc[0] | 0,
            mapPointerY: cc[1] | 0,
        });
    }

    protected externalInteracting = false;
    protected viewChangedFlag = false;

    setInteracting(on) {
        this.externalInteracting = on
    }

    rawUpdateCount = 0;

    protected viewChanged(arg) {
        if (this.connectedToStore) {
            this.update({
                mapRawUpdateCount: ++this.rawUpdateCount,
            })
        }

        this.viewChangedFlag = true
    }

    watchViewChanges() {
        if (this.externalInteracting || this.oView.getInteracting() || this.oView.getAnimating()) {
            return
        }

        if (this.viewChangedFlag) {
            this.updateViewState();
            this.updateAppLocation();
            this.changed();
            this.viewChangedFlag = false;
        }
    }

    protected initEvents() {

        let vs = this.viewState;

        setInterval(() => this.watchViewChanges(), VIEW_UPDATE_INTERVAL)

        this.updatePointerPosition([vs.centerX, vs.centerY]);
        this.viewChanged('init');

        this.oMap.on('pointermove', tools.debounce(
            e => this.updatePointerPosition(e.coordinate),
            POINTER_UPDATE_INTERVAL));

        this.oView.on('change:resolution', () => this.viewChanged('res'));
        this.oView.on('change:center', () => this.viewChanged('center'));
        this.oView.on('change:rotation', () => this.viewChanged('rotation'));
    }

    protected initInteractions() {
        this.standardInteractions = {
            'DragPan': new ol.interaction.DragPan({
                kinetic: new ol.Kinetic(-0.01, 0.1, 100)

            }),
            'PinchZoom': new ol.interaction.PinchZoom({
                constrainResolution: true
            }),
            'MouseWheelZoom': new ol.interaction.MouseWheelZoom({
                constrainResolution: true
            }),
            'ZoomBox': new ol.interaction.DragZoom({
                condition: ol.events.condition.shiftKeyOnly,
                className: 'modZoomBox',
            })
        }
    }

    //

    newFeature(args: types.IMapFeatureArgs) {
        return new Feature(this, args);
    }

    readFeature(props) {
        return new Feature(this, {props});
    }

    readFeatures(fs) {
        return fs.map(props => new Feature(this, {props}));
    }

    writeFeatures(fs) {
        return fs.map(f => f.props);
    }

    //

    geom2shape(geom) {
        if (!geom)
            return null;

        // NB: this is our geojson extension
        // the server is supposed to deal with this in a reasonable way
        if (geom instanceof ol.geom.Circle) {
            return {
                crs: this.crs,
                geometry: {
                    type: 'Circle',
                    center: (geom as ol.geom.Circle).getCenter(),
                    radius: (geom as ol.geom.Circle).getRadius(),
                }
            }
        }

        let format = new ol.format.GeoJSON({
            defaultDataProjection: this.projection,
            featureProjection: this.projection,
        });
        return {
            crs: this.crs,
            // @TODO: silly
            geometry: JSON.parse(format.writeGeometry(geom))
        }
    }

    shape2geom(shape) {
        if (!shape || !shape.geometry)
            return null;

        // @TODO: shapes are assumed to be in the map projection

        if (shape.geometry.type === 'Circle') {
            return new ol.geom.Circle(
                shape.geometry.center,
                shape.geometry.radius,
            );
        }

        let format = new ol.format.GeoJSON();
        return format.readGeometry(shape.geometry);
    }

    //

    async searchForFeatures(args) {

        let ls = this.searchLayers();
        let params: api.SearchParams = {
            bbox: this.bbox,
            keyword: args.keyword || '',
            layerUids: tools.compact(ls.map(la => la.uid)),
            resolution: this.viewState.resolution,
            limit: args.limit || 999
        };

        if (args.geometry) {
            params.shapes = [this.geom2shape(args.geometry)]
        }

        let res = await this.app.server.searchFindFeatures(params);

        if (res.error) {
            console.log('SEARCH_ERROR', res);
            return [];
        }

        let features = this.readFeatures(res.features);
        return features;
    }

    protected searchLayers() {
        // select visible terminal layers, starting with selected layers, if any
        // otherwise with the root
        let roots = this.collect(this.root, la => la.selected);
        if (!roots.length)
            roots = [this.root];

        let layers = [];

        roots.forEach(root => this.walk(root, layer => {
            if (!layer.shouldDraw)
                return MapManager.STOP_WALK;
            if (!layer.isSystem)
                layers.push(layer)
        }));

        return tools.uniq(layers);
    }

    protected async printItems(boxRect, dpi): Promise<Array<api.PrintItem>> {
        let _this = this;
        let items: Array<api.PrintItem> = [];

        function makeBitmap2(): api.PrintItem {
            let canvas = _this.oMap.getViewport().firstChild as HTMLCanvasElement;

            let rc = canvas.getBoundingClientRect(),
                rb = boxRect,
                cx = rb.left - rc.left,
                cy = rb.top - rc.top;

            let ctx = canvas.getContext('2d'),
                imgData = ctx.getImageData(cx, cy, rb.width, rb.height);

            if (USE_RAW_BITMAPS_FOR_PRINT) {
                return {
                    type: 'bitmap',
                    mode: 'RGBA',
                    data: imgData.data,
                    width: imgData.width,
                    height: imgData.height,
                };
            }

            let cnv2 = document.createElement('canvas');

            cnv2.width = rb.width;
            cnv2.height = rb.height;

            cnv2.getContext('2d').putImageData(imgData, 0, 0);

            return {
                type: 'url',
                url: cnv2.toDataURL()
            };
        }

        async function makeBitmap(layers): Promise<api.PrintItem> {
            let hidden = [];

            _this.walk(_this.root, la => {
                if (la.visible && !layers.includes(la)) {
                    hidden.push(la);
                    la.hide();
                }
            });

            let bmp: api.PrintItem;

            await tools.delay(200, () => {
                console.time('creating_bitmap');
                bmp = makeBitmap2();
                console.timeEnd('creating_bitmap');
            });

            hidden.forEach(la => la.show());

            return bmp;
        }

        if (boxRect && dpi === 0) {
            // draft printing, print everything as a bitmap

            let bmpLayers = [];

            this.walk(this.root, la => {
                let pi = la.shouldDraw && la.printItem;
                if (pi) {
                    bmpLayers.push(la);
                }
            });

            items.push(await makeBitmap(bmpLayers));
            return items;
        }

        if (boxRect && dpi <= BITMAP_PRINT_DPI_THRESHOLD) {
            // low-res printing, print rasters as bitmaps

            let bmpLayers = [];

            this.walk(this.root, la => {
                let pi = la.shouldDraw && la.printItem;
                if (pi) {
                    if (pi.type === 'raster') {
                        items.push(null);
                        bmpLayers.push(la);
                    } else {
                        items.push(pi)
                    }
                }
            });

            for (let i = 0; i < items.length; i++) {
                if (!items[i]) {
                    items[i] = await makeBitmap([bmpLayers.shift()]);
                }
            }

            return items;
        }

        // normal printing, pass each layer's printItem as is

        this.walk(this.root, la => {
            let pi = la.shouldDraw && la.printItem;
            if (pi) {
                items.push(pi);
            }
        });

        return items;
    }

    async basicPrintParams(boxRect, dpi) {
        let vs = this.viewState;

        return {
            items: await this.printItems(boxRect, dpi),
            rotation: Math.round(tools.rad2deg(vs.rotation)),
            scale: tools.res2scale(vs.resolution),
        }
    }

    //

    protected visibleAttributions() {
        let a = [this.app.project.meta.attribution];

        this.walk(this.root, (layer: types.IMapLayer) => {
            if (layer.shouldDraw)
                a.push(layer.attribution);
        });

        return tools.uniq(tools.compact(a));
    }

    protected insertLayer(layer, where, parent) {
        if (parent) {
            layer.parent = parent;
            if (where < 0)
                parent.children.push(layer);
            else
                parent.children.splice(where, 0, layer);
        } else {
            layer.parent = null;
        }
    }

    protected draw() {
        this.walk(this.root, la => la.beforeDraw());

        let layers = [];

        this.walk(this.root, layer => {
            if (!layer.shouldDraw)
                return MapManager.STOP_WALK;
            if (layer.oLayer)
                layers.push(layer)
        });

        this.oMap.setLayerGroup(new ol.layer.Group({
            layers: layers.map(la => la.oLayer).reverse()
        }));

        this.walk(this.root, la => {
            if (layers.indexOf(la) < 0)
                this.app.server.dequeueLoad(la.uid);
        });
    }

    walk(layer, fn: (la: types.IMapLayer) => any) {
        if (fn(layer) === MapManager.STOP_WALK)
            return;
        layer.children.forEach(la => this.walk(la, fn));
    }

    collect(layer, fn) {
        let ls = [];

        let walk = layer => {
            let r = fn(layer);
            if (r === MapManager.STOP_WALK)
                return;
            if (r)
                ls.push(layer);
            layer.children.forEach(walk);
        };

        walk(layer);
        return ls;
    }

}
