// gws/map/model.ts
import * as types from '../types';
import * as api from '../core/gws-api';
import * as tools from '../tools';


export class ModelRegistry implements types.IModelRegistry {
    models: { [uid: string]: Model };
    map: types.IMapManager;

    constructor(map: types.IMapManager) {
        this.map = map;
    }


    setProps(props: Array<api.ModelProps>): ModelRegistry {
        this.models = {};

        for (let p of props) {
            this.models[p.uid] = new Model(this);
        }

        for (let p of props) {
            this.models[p.uid].setProps(p);
        }
        return this;
    }

    getModel(uid) {
        return this.models[uid];
    }

    getModelForLayer(layer) {
        for (let m of Object.values(this.models))
            if (m.layerUid === layer.uid)
                return m
    }

}

export class Model implements types.IModel {
    fields: Array<types.IModelField>;
    geometryName: string;
    keyName: string;
    layerUid: string;
    uid: string;
    registry: ModelRegistry;

    constructor(registry) {
        this.registry = registry;
    }

    setProps(props: api.ModelProps): Model {
        this.geometryName = props.geometryName;
        this.keyName = props.keyName;
        this.layerUid = props.layerUid;
        this.uid = props.uid;

        this.fields = [];

        if (props['fields'])
            for (let p of props['fields'])
                this.fields.push(new ModelField(this).setProps(p));

        return this;
    }

    getLayer() {
        return this.registry.map.getLayer(this.layerUid) as types.IFeatureLayer;
    }

    getField(name) {
        for (let f of this.fields)
            if (f.name === name)
                return f;
    }

}

export class ModelField implements types.IModelField {
    name: string;
    type: string;
    dataType: string;
    geometryType: string;
    title: string;

    relations: Array<types.IModelRelation>;
    widget?: api.ModelWidgetProps;

    model: Model;

    constructor(model) {
        this.model = model;

    }

    setProps(props: api.ModelFieldProps): ModelField {
        this.name = props.name;
        this.dataType = props.dataType;
        this.geometryType = props.geometryType;
        this.title = props.title;
        this.widget = props.widget;


        this.relations = [];

        if (props.relations) {
            for (let r of props.relations) {
                this.relations.push({
                    type: r.type,
                    model: this.model.registry.getModel(r.modelUid),
                    fieldName: r.fieldName,
                    title: r.title,
                })
            }
        }

        return this;
    }
}
