// gws/app/root-controller.tsx
import * as React from 'react';

import {ViewProps} from '../types';
import {View} from './view';
import {Controller} from './controller';

import * as tools from '../tools';

interface RootViewProps extends ViewProps {
    controller: RootController;
}

class RootView extends View<RootViewProps> {
    render() {
        return this.props.controller.app.store.wrap(
            this.props.controller.appView()
        )
    }
}

interface AppViewProps extends ViewProps {
    controller: RootController;
}

class AppView extends View<AppViewProps> {
    componentDidMount() {
        this.props.controller.app.whenMounted();
    }

    render() {

        let views = (name) => this.props.controller.children.map(cc =>
            <React.Fragment key={cc.uid + name}>{cc[name]}</React.Fragment>
        );

        return <div>
            <div className="gwsMap"/>
            <div className="gwsMapOverlay">{views('mapOverlayView')}</div>
            {views('defaultView')}
            <div className="gwsAppOverlay">{views('appOverlayView')}</div>
        </div>
    }
}

export class RootController extends Controller {
    appClasses = [];

    protected setClass(yes, cls) {
        this.appClasses = this.appClasses.filter(x => x !== cls);
        if (yes)
            this.appClasses.push(cls);
        this.app.domNode.className = this.appClasses.join(' ');
    }

    async init() {
        await super.init();
        this.appClasses = this.app.domNode.className.split(' ');

        this.whenChanged('sidebarVisible', v => this.setClass(v, 'withSidebar'));
        this.whenChanged('printState', v => this.setClass(v, 'withPrintPreview'));
    }

    appView() {
        let c = this.connect(AppView);
        return this.createElement(c, {children: this.children});
    }

    get defaultView() {
        return this.createElement(RootView, {children: this.children});
    }
}
