// gws/core/server.ts
import axios from 'axios';
import * as base from './gws-server';

function errorResponse(err) {
    console.log('AXIOS ERROR', err);
    return {
        error: {
            status: err.response ? err.response.status : -1,
            info: String(err)
        }
    }
}

export class Server extends base.GwsServer {
    url: string;
    whenRequestStarted: any = null;
    whenRequestEnded: any = null;

    constructor(app, url) {
        super();
        this.url = url;
    }

    async _call(cmd, params) {
        this.whenRequestStarted && this.whenRequestStarted(cmd, params);
        let res = await this._call2(cmd, params);
        this.whenRequestEnded && this.whenRequestEnded(cmd, params, res);
        return res;
    }

    async _call2(cmd, params) {
        try {
            return await axios.post(this.url, {cmd, params}, {withCredentials: true})
                .then(res => res.data)
                .catch(err => errorResponse(err));
        } catch (err) {
            return await errorResponse(err);
        }
    }
}
