// gws/core/server.ts
import axios, {Canceler, ResponseType} from 'axios';
import * as msgpack from '@ygoe/msgpack';
import * as api from './gws-api';

function errorResponse(err) {
    console.log('AXIOS ERROR', err);
    return {
        error: {
            status: err.response ? err.response.status : -1,
            info: String(err)
        }
    }
}

enum QueueState {
    OPEN = 1,
    RUNNING = 2,
    DONE = 3
}

class QueuedRequest {
    layerUid: string;
    url: string;
    responseType: ResponseType;
    state: QueueState;
    server: Server;
    cancel: Canceler;
    resolve: any;
    promise: Promise<any>;

    constructor(server: Server, layerUid, url, responseType) {
        this.server = server;
        this.layerUid = layerUid;
        this.url = url;
        this.responseType = responseType;
        this.state = QueueState.OPEN;

        this.promise = new Promise<any>(resolve => this.resolve = resolve);
    }

    abort() {
        // console.log('SERVER.queue.ABORT', this.layerUid);
        if (this.cancel) {
            this.cancel();
        }
        this.end(null);
    }

    end(data) {
        this.state = QueueState.DONE;
        if (this.resolve) {
            this.resolve(data);
        }
        this.resolve = null;
    }

    start() {
        // console.log('SERVER.queue.START', this.layerUid);

        this.state = QueueState.RUNNING;

        axios.get(this.url, {
            cancelToken: new axios.CancelToken((c) => this.cancel = c),
            responseType: this.responseType
        })
            .then(res => {
                // console.log('SERVER.queue.OK', this.layerUid, res.data.size);
                this.end(res.data)
            })
            .catch(err => {
                // console.log('SERVER.queue.ERROR', this.layerUid, err);
                this.end(null)
            });
    }

}

const MSGPACK_MIME = 'application/msgpack';

export class Server extends api.GwsServer {
    url: string;
    whenChanged: () => void;
    commandCount = 0;
    qOpen: Array<QueuedRequest> = [];
    qRun: Array<QueuedRequest> = [];
    qSize = 5;

    protected app;

    constructor(app, url) {
        super();
        this.url = url;
        this.whenChanged = () => null;
        this.app = app;
        window.setInterval(() => this.pollQueue(), 500);
    }

    get requestCount() {
        return this.commandCount + this.qRun.length + this.qOpen.length;
    }

    async getUrl(url: string): Promise<string> {
        try {
            let res = await axios.request({url});
            return res.data;
        } catch (err) {
            return '';
        }
    }

    async _call(cmd, params, options) {
        params = params || {};
        params.projectUid = params.projectUid || this.app.project.uid;
        params.localeUid = params.localeUid || this.app.localeUid;

        this.commandCount++;
        this.whenChanged();
        let res = await this._call2(cmd, params, options);
        this.commandCount--;
        this.whenChanged();
        return res;
    }

    async _call2(cmd, params, options) {
        let req: any = {
            url: this.url,
            method: 'POST',
            data: {cmd, params},
            withCredentials: true
        };

        if (options && options.binary) {
            req.data = msgpack.serialize(req.data);
            req.headers = {
                'content-type': MSGPACK_MIME,
                'accept': MSGPACK_MIME
            };
            req.responseType = 'arraybuffer';
        }

        try {
            let res = await axios.request(req),
                data = res.data;
            if (res.headers['content-type'].includes(MSGPACK_MIME))
                data = msgpack.deserialize(new Uint8Array(data));
            return data;
        } catch (err) {
            if (err && err.response && err.response.data)
                return err.response.data;
            return errorResponse(err);
        }
    }

    queueLoad(layerUid, url, responseType) {
        this.dequeueLoad(layerUid);

        if (!url)
            return;

        let req = new QueuedRequest(this, layerUid, url, responseType);
        this.qOpen.push(req);

        // console.log('SERVER.queue.ADD', layerUid);

        return req.promise;
    }

    dequeueLoad(layerUid) {
        // console.log('SERVER.queue.REMOVE', layerUid);
        this.qOpen = this.qOpen.filter(r => r.layerUid !== layerUid);
        this.qRun.forEach(r => {
            if (r.layerUid === layerUid) {
                r.abort();
            }
        });
    }

    pollQueue() {
        let len = this.qRun.length,
            changed = 0;

        this.qRun = this.qRun.filter(r => r.state !== QueueState.DONE);
        if (this.qRun.length !== len)
            changed++;

        if (this.qRun.length < this.qSize && this.qOpen.length) {
            let r = this.qOpen.shift()
            this.qRun.push(r)
            r.start();
            changed++;
        }

        if (changed) {
            this.whenChanged();
            // console.log('LOAD:POLL', this.qRun.length, '/', this.qOpen.length)
        }
    }
}
