// gws/core/server.ts
import axios, {Canceler} from 'axios';
import * as api from './gws-api';

function errorResponse(err) {
    console.log('AXIOS ERROR', err);
    return {
        error: {
            status: err.response ? err.response.status : -1,
            info: String(err)
        }
    }
}

enum QueueState {
    OPEN = 1,
    RUNNING = 2,
    DONE = 3
}

class QueuedRequest {
    layerUid: string;
    url: string;
    responseType: string;
    state: QueueState;
    server: Server;
    cancel: Canceler;
    resolve: any;
    promise: Promise<any>;

    constructor(server: Server, layerUid, url, responseType) {
        this.server = server;
        this.layerUid = layerUid;
        this.url = url;
        this.responseType = responseType;
        this.state = QueueState.OPEN;

        this.promise = new Promise<any>(resolve => this.resolve = resolve);
    }

    abort() {
        console.log('ABORT', this.layerUid)
        if (this.cancel) {
            this.cancel();
        }
        this.end(null);
    }

    end(data) {
        this.state = QueueState.DONE;
        if (this.resolve) {
            this.resolve(data);
        }
        this.resolve = null;
    }

    start() {
        console.log('START', this.layerUid);

        this.state = QueueState.RUNNING;

        axios.get(this.url, {
            cancelToken: new axios.CancelToken((c) => this.cancel = c),
            responseType: this.responseType
        })
            .then(res => {
                console.log('OK', this.layerUid);
                this.end(res.data)
            })
            .catch(err => {
                console.log('ERR', this.layerUid, err);
                this.end(null)
            });
    }

}

export class Server extends api.GwsServer {
    url: string;
    whenChanged: () => void;
    commandCount = 0;
    queue: Array<QueuedRequest> = [];
    queueSize = 3;

    constructor(app, url) {
        super();
        this.url = url;
        this.whenChanged = () => null;
        window.setInterval(() => this.pollQueue(), 500);
    }

    get requestCount() {
        return this.commandCount + this.queue.length;
    }

    async _call(cmd, params) {
        this.commandCount++;
        this.whenChanged();
        let res = await this._call2(cmd, params);
        this.commandCount--;
        this.whenChanged();
        return res;
    }

    async _call2(cmd, params) {
        try {
            return await axios.post(this.url, {cmd, params}, {withCredentials: true})
                .then(res => res.data)
                .catch(err => errorResponse(err));
        } catch (err) {
            return await errorResponse(err);
        }
    }

    queueLoad(layerUid, url, responseType) {
        this.dequeueLoad(layerUid);

        if (!url)
            return;

        let req = new QueuedRequest(this, layerUid, url, responseType);
        this.queue.push(req);

        console.log('QUEUE', layerUid);

        return req.promise;
    }

    dequeueLoad(layerUid) {
        this.queue = this.queue.filter(r => r.state !== QueueState.DONE);

        this.queue.forEach(r => {
            if (r.layerUid === layerUid)
                r.abort();
        });
    }

    prevLen = 0;

    pollQueue() {
        if (!this.queue.length && !this.prevLen)
            return;

        this.queue = this.queue.filter(r => r.state !== QueueState.DONE);
        this.prevLen = this.queue.length;

        if (this.queue.length === 0) {
            console.log('EMPTY');
            this.whenChanged();
            return;
        }

        let running = this.queue.filter(r => r.state === QueueState.RUNNING).length;
        let req = this.queue.find(r => r.state === QueueState.OPEN);

        console.log('POLL', running, '/', this.queue.length, this.queue.map(r => r.layerUid))

        if (running < this.queueSize && req) {
            req.start();
            this.whenChanged();
        }
    }
}
