// gws/components/feature.tsx
import * as React from 'react';

import * as types from '../types';
import * as ui from '../ui';
import * as tools from '../tools';
import * as list from './list';

let {Row, Cell} = ui.Layout;

interface BaseListProps extends types.ViewProps {
    controller: types.IController;
    features: Array<types.IMapFeature>;
    content?: (f: types.IMapFeature) => React.ReactNode;
    isSelected?: (f: types.IMapFeature) => boolean;
}

interface ListProps extends BaseListProps {
    leftButton?: (f: types.IMapFeature) => React.ReactNode;
    rightButton?: (f: types.IMapFeature) => React.ReactNode;
    withZoom?: boolean;
}

interface InfoListProps extends BaseListProps {
}

interface InfoListState {
    selectedIndex?: number;
}

class FeatureList extends list.List<types.IMapFeature> {
}

export class List extends React.PureComponent<ListProps> {
    render() {
        let zoom = f => this.props.controller.update({
            marker: {
                features: [f],
                mode: 'zoom draw fade'
            }
        });

        return <FeatureList
            controller={this.props.controller}
            items={this.props.features}
            content={f => this.props.content ? this.props.content(f) : <ui.Text content={f.props.title}/>}
            uid={f => f.uid}
            isSelected={this.props.isSelected}
            leftButton={this.props.withZoom
                ? f => <list.Button className="cmpListZoomListButton" whenTouched={() => zoom(f)}/>
                : this.props.leftButton}
            rightButton={this.props.rightButton}
        />;
    }
}

export class InfoList extends React.Component<InfoListProps, InfoListState> {

    state = {selectedIndex: 0};

    componentDidMount() {
        this.setState({selectedIndex: 0})
    }

    componentDidUpdate(prevProps: InfoListProps) {
        if (prevProps.features !== this.props.features) {
            console.log('NEW FEATURES');
            this.setState({selectedIndex: 0})
        }
    }

    show(n) {
        this.setState({
            selectedIndex: n
        });
        this.props.controller.update({
            marker: {
                features: [this.props.features[n]],
                mode: 'draw',
            }
        });
    }

    render() {
        let cc = this.props.controller;
        let sel = this.state.selectedIndex || 0;
        let f = this.props.features[sel];

        if (!f)
            return null;

        let len = this.props.features.length;

        let dec = () => this.show(sel === 0 ? len - 1 : sel - 1);
        let inc = () => this.show(sel === len - 1 ? 0 : sel + 1);

        let zoom = f => cc.update({
            marker: {
                features: [f],
                mode: 'zoom draw',
            }
        });

        let close = () => cc.update({
            marker: null,
            infoboxContent: null
        });

        let lens = f => {
            cc.update({
                lensGeometry: f.geometry
            });
            cc.app.startTool('Tool.Lens')
        };

        let select = f => cc.update({
            selectAddFeature: f
        });

        // if(!cc.app.controllerByTag('Toolbar.Lens'))
        //     lens = null;
        //
        // if(!cc.app.controllerByTag('Toolbar.Select'))
        //     select = null;

        let item = this.props.content;
        if (!item)
            item = f => <ui.TextBlock className="cmpDescription" withHTML content={f.props.description}/>;

        return <div className="modInfoboxContent">
            <div className="modInfoboxBody">
                {item(f)}
            </div>
            <div className="modInfoboxFooter">
                <Row>
                    <Cell flex/>

                    {len > 1 && <Cell>
                        <div className="modInfoboxPagerText">
                            {this.state.selectedIndex + 1} / {len}
                        </div>
                    </Cell>}
                    {len > 1 && <Cell>
                        <ui.IconButton
                            className='modInfoboxPagerBack'
                            tooltip={cc.__('modInfoboxPagerBack')}
                            whenTouched={dec}/>
                    </Cell>}
                    {len > 1 && <Cell>
                        <ui.IconButton
                            className='modInfoboxPagerForward'
                            tooltip={cc.__('modInfoboxPagerForward')}
                            whenTouched={inc}/>
                    </Cell>}

                    {lens && <Cell>
                        <ui.IconButton
                            className='modInfoboxLensButton'
                            tooltip={cc.__('modInfoboxLensButton')}
                            whenTouched={() => lens(f)}
                        />
                    </Cell>}

                    {select && <Cell>
                        <ui.IconButton
                            className='modInfoboxSelectButton'
                            tooltip={cc.__('modInfoboxSelectButton')}
                            whenTouched={() => select(f)}
                        />
                    </Cell>}

                    <Cell>
                        <ui.IconButton
                            className='modInfoboxZoomButton'
                            tooltip={cc.__('modInfoboxZoomButton')}
                            whenTouched={() => zoom(f)}
                        />
                    </Cell>
                    <Cell>
                        <ui.IconButton
                            className='modInfoboxCloseButton'
                            tooltip={cc.__('modInfoboxCloseButton')}
                            whenTouched={close}
                        />
                    </Cell>
                </Row>
            </div>
        </div>
    }
}
