// gws/components/feature.tsx
import * as React from 'react';

import * as types from '../types';
import * as ui from '../ui';
import * as tools from '../tools';

let {Row, Cell} = ui.Layout;

interface BaseListProps {
    controller: types.IController;
    features: Array<types.IMapFeature>;
    item?: (f: types.IMapFeature) => React.ReactNode;
    isSelected?: (f: types.IMapFeature) => boolean;
}

interface ListProps extends BaseListProps {
    leftIcon?: (f: types.IMapFeature) => React.ReactNode;
    rightIcon?: (f: types.IMapFeature) => React.ReactNode;
}

interface ListItemProps extends ListProps {
    feature: types.IMapFeature;
}

interface PopupListProps extends BaseListProps {
}

interface PopupListState {
    selectedIndex?: number;
}

export class List extends React.PureComponent<ListProps> {
    render() {
        return <div className="cmpFeatureList">
            {this.props.features.map(f =>
                <ListItem key={f.uid} feature={f} {...this.props} />)}
        </div>
    }
}

class ListItem extends React.PureComponent<ListItemProps> {
    render() {
        let f = this.props.feature,
            item = this.props.item
                ? this.props.item(f)
                : <ui.Text content={f.props.title}/>,
            leftIcon = this.props.leftIcon
                ? this.props.leftIcon(f)
                : null,
            rightIcon = this.props.rightIcon
                ? this.props.rightIcon(f)
                : null,
            selected = this.props.isSelected && this.props.isSelected(f);

        return <Row {...tools.cls('cmpFeatureListItem', selected && 'isSelected')}>
            {leftIcon && <Cell className="cmpFeatureListIcon">{leftIcon}</Cell>}
            <Cell flex className="cmpFeatureListContent">{item}</Cell>
            {rightIcon && <Cell className="cmpFeatureListIcon">{rightIcon}</Cell>}
        </Row>
    }
}

export class PopupList extends React.Component<PopupListProps, PopupListState> {

    state = {selectedIndex: 0};

    componentDidMount() {
        this.setState({selectedIndex: 0})
    }

    componentDidUpdate(prevProps: PopupListProps) {
        if (prevProps.features !== this.props.features) {
            console.log('NEW FEATURES');
            this.setState({selectedIndex: 0})
        }
    }

    show(n) {
        this.setState({
            selectedIndex: n
        });
        this.props.controller.update({
            marker: {
                features: [this.props.features[n]],
                mode: 'draw',
            }
        });
    }

    render() {
        let sel = this.state.selectedIndex || 0;
        let f = this.props.features[sel];

        if (!f)
            return null;

        let len = this.props.features.length;

        let dec = () => this.show(sel === 0 ? len - 1 : sel - 1);
        let inc = () => this.show(sel === len - 1 ? 0 : sel + 1);

        let zoom = f => {
            this.props.controller.update({
                marker: {
                    features: [f],
                    mode: 'zoom draw',
                }
            });
        };

        let close = () => this.props.controller.update({
            marker: null,
            popupContent: null
        });

        let item = this.props.item;
        if (!item)
            item = f => <ui.TextBlock className="cmpDescription" withHTML content={f.props.description}/>;

        return <div className="modPopupContent">
            <div className="modPopupBody">
                {item(f)}
            </div>
            <div className="modPopupFooter">
                <Row>
                    <Cell flex/>

                    {len > 1 && <Cell>
                        <div className="modPopupPagerText">
                            {this.state.selectedIndex + 1} / {len}
                        </div>
                    </Cell>}
                    {len > 1 && <Cell>
                        <ui.IconButton
                            className='modPopupPagerBack'
                            tooltip={this.props.controller.__('modPopupPagerBack')}
                            whenTouched={dec}/>
                    </Cell>}
                    {len > 1 && <Cell>
                        <ui.IconButton
                            className='modPopupPagerForward'
                            tooltip={this.props.controller.__('modPopupPagerForward')}
                            whenTouched={inc}/>
                    </Cell>}


                    <Cell>
                        <ui.IconButton
                            className='modPopupZoomButton'
                            tooltip={this.props.controller.__('modPopupZoomButton')}
                            whenTouched={() => zoom(f)}
                        />
                    </Cell>
                    <Cell>
                        <ui.IconButton
                            className='modPopupCloseButton'
                            tooltip={this.props.controller.__('modPopupCloseButton')}
                            whenTouched={close}
                        />
                    </Cell>
                </Row>
            </div>
        </div>
    }
}
