// gws/components/form.tsx
import * as React from 'react';

import * as types from '../types';
import * as api from '../core/gws-api';
import * as ui from '../ui';
import * as tools from '../tools';
import * as featureComp from './feature';
import * as listComp from './list';
import * as widget from './widget';

let {Row, Cell} = ui.Layout;


interface FormProps {
    controller: types.IController;
    model: types.IModel;
    feature: types.IFeature;
    values: types.Dict;
    errors?: types.Dict;
    makeWidget: (field: types.IModelField, feature: types.IFeature, values: types.Dict) => React.ReactElement | null;
}

interface FormFieldProps {
    controller: types.IController;
    field: types.IModelField;
    feature: types.IFeature;
    values: types.Dict;
    errors?: types.Dict;
    makeWidget: (field: types.IModelField, feature: types.IFeature, values: types.Dict) => React.ReactElement | null;
}


//

//


export class FormField extends React.PureComponent<FormFieldProps> {
    render() {
        let field = this.props.field;

        let widget = this.props.makeWidget(field, this.props.feature, this.props.values);
        if (!widget)
            return null;

        let err = this.props.errors ? this.props.errors[field.name] : null;

        return <React.Fragment key={field.name}>
            <tr className={err ? 'isError' : ''}>
                <th>
                    {field.title}
                </th>
                <td>
                    {widget}
                </td>
            </tr>
            <tr {...tools.cls('cmpFormError', err && 'isActive')}>
                <th>&nbsp;</th>
                <td>{err}</td>
            </tr>
        </React.Fragment>

    }

}

export class Form extends React.PureComponent<FormProps> {


    render() {
        return <table className="cmpForm">
            <tbody>
            {this.props.model.fields.map(f => <FormField
                field={f}
                controller={this.props.controller}
                feature={this.props.feature}
                values={this.props.values}
                makeWidget={this.props.makeWidget}
                errors={this.props.errors}
            />)}
            </tbody>
        </table>
    }
}
