// gws/components/list.tsx
import * as React from 'react';

import * as types from '../types';
import * as ui from '../ui';
import * as tools from '../tools';

let {Row, Cell} = ui.Layout;

interface ListProps extends types.ViewProps {
    controller: types.IController;
    items: Array<any>;
    content?: (it) => React.ReactNode;
    uid?: (it) => string;
    isSelected?: (f: any) => boolean;
    leftButton?: (f: any) => React.ReactNode;
    rightButton?: (f: any) => React.ReactNode;
}

interface ListItemProps extends ListProps {
    item: any;
}

interface ListButtonProps {
    className: string;
    tooltip?: string;
    whenTouched: () => void;
}

class ListItem extends React.PureComponent<ListItemProps> {
    render() {
        let it = this.props.item,
            content = this.props.content ? this.props.content(it) : <span>String(it)</span>,
            leftButton = this.props.leftButton && this.props.leftButton(it),
            rightButton = this.props.rightButton && this.props.rightButton(it),
            selected = this.props.isSelected && this.props.isSelected(it);

        return <Row {...tools.cls('cmpListRow', selected && 'isSelected')}>
            {leftButton && <Cell>{leftButton}</Cell>}
            <Cell flex className="cmpListContent">{content}</Cell>
            {rightButton && <Cell>{rightButton}</Cell>}
        </Row>
    }
}

export class List extends React.PureComponent<ListProps> {
    render() {
        return <div className="cmpList">
            {this.props.items.map((it, n) => <ListItem
                    {...this.props}
                    key={this.props.uid ? this.props.uid(it) : String(n)}
                    item={it}
                />
            )}
        </div>
    }
}

export class Button extends React.PureComponent<ListButtonProps> {
    render() {
        return <ui.Button {...this.props} />;
    }
}
