// gws/components/list.tsx
import * as React from 'react';

import * as types from '../types';
import * as ui from '../ui';
import * as tools from '../tools';

let {Row, Cell} = ui.Layout;

interface ListProps<T> extends types.ViewProps {
    controller: types.IController;
    items: Array<T>;
    content: (it: T) => React.ReactNode;
    uid?: (it: T) => string;
    isSelected?: (f: any) => boolean;
    leftButton?: (f: any) => React.ReactNode;
    rightButton?: (f: any) => React.ReactNode;
}

interface ListItemProps<T> extends ListProps<T> {
    it: T;
}

interface ListButtonProps {
    className: string;
    tooltip?: string;
    whenTouched: () => void;
}

class ListItem<T> extends React.PureComponent<ListItemProps<T>> {
    render() {
        let it = this.props.it,
            content = this.props.content(it),
            leftButton = this.props.leftButton && this.props.leftButton(it),
            rightButton = this.props.rightButton && this.props.rightButton(it),
            selected = this.props.isSelected && this.props.isSelected(it);

        return <Row {...tools.cls('cmpListButton', selected && 'isSelected')}>
            {leftButton && <Cell>{leftButton}</Cell>}
            <Cell flex className="cmpListContent">{content}</Cell>
            {rightButton && <Cell className="cmpListButton">{rightButton}</Cell>}
        </Row>
    }
}

export class List<T> extends React.PureComponent<ListProps<T>> {
    render() {
        return <div className="cmpList">
            {this.props.items.map((it, n) => <ListItem
                    {...this.props}
                    key={this.props.uid ? this.props.uid(it) : String(n)}
                    it={it}
                />
            )}
        </div>
    }
}

export class Button extends React.PureComponent<ListButtonProps> {
    render() {
        return <ui.Button {...this.props} />;
    }
}
